aten/src/ATen/native/cuda/IndexKernel.cu

```diff
@@ -5,15 +5,67 @@
 #include <ATen/native/TensorIterator.h>
 #include <ATen/core/Array.h>
 #include <ATen/cuda/detail/OffsetCalculator.cuh>
+#include <ATen/cuda/detail/IndexUtils.cuh>
 #include <ATen/cuda/CUDAContext.h>
+#include <ATen/cuda/CUDAApplyUtils.cuh>
 #include <ATen/ExpandUtils.h>
+#include <THC/THCTensorInfo.cuh>
 
 namespace at { namespace native {
 
 static constexpr int launch_bound2 = 4;
 
 static constexpr int launch_size_nd = 128;
 
+template <int Dims, typename T, typename IndexType>
+__device__ __forceinline__ IndexType indexToOffset(
+    const cuda::detail::TensorInfo<T, IndexType>& info,
+    int64_t index,
+    IndexType size)
+{
+  IndexType linearIndex = static_cast<IndexType>(index);
+  CUDA_KERNEL_ASSERT(linearIndex < size && linearIndex >= -size);
+  if (linearIndex < 0) {
+    linearIndex += size;
+  }
+  return cuda::detail::IndexToOffset<T, IndexType, Dims>::get(linearIndex, info);
+}
+
+template<typename IndexType, typename T>
+void dispatchTakePutImpl(const Tensor& input, Tensor& output, const Tensor& index) {
+  auto inputInfo = cuda::detail::getTensorInfo<T, IndexType>(input);
+  inputInfo.collapseDims();
+  auto numel = input.numel();
+  if (inputInfo.isContiguous()) {
+    cuda::CUDA_tensor_apply2<T, int64_t>(
+        output,
+        index,
+        [inputInfo, numel] __device__ (
```

Comment:
change: Passing in a lambda instead of a `struct TensorTakeOp` object (see the deleted code in `THCTensorIndex.cu`). I assumed that the lack of lambdas was because of the fact that TH is a legacy api.